File: app/src/main/java/org/wikipedia/feed/aggregated/AggregatedFeedContentClient.java
package org.wikipedia.feed.aggregated;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.util.Pair;

import org.wikipedia.WikipediaApp;
import org.wikipedia.dataclient.ServiceFactory;
import org.wikipedia.dataclient.WikiSite;
import org.wikipedia.feed.FeedContentType;
import org.wikipedia.feed.FeedCoordinator;
import org.wikipedia.feed.dataclient.FeedClient;
import org.wikipedia.feed.featured.FeaturedArticleCard;
import org.wikipedia.feed.image.FeaturedImageCard;
import org.wikipedia.feed.model.Card;
import org.wikipedia.feed.model.UtcDate;
import org.wikipedia.feed.mostread.MostReadListCard;
import org.wikipedia.feed.news.NewsCard;
import org.wikipedia.feed.onthisday.OnThisDayCard;
import org.wikipedia.util.DateUtil;
import org.wikipedia.util.log.L;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class AggregatedFeedContentClient {
    @NonNull private Map<String, AggregatedFeedContent> aggregatedResponses = new HashMap<>();
    private int aggregatedResponseAge = -1;
    private CompositeDisposable disposables = new CompositeDisposable();

    public static class OnThisDayFeed extends BaseClient {
        public OnThisDayFeed(@NonNull AggregatedFeedContentClient aggregatedClient) {
            super(aggregatedClient);
        }

        @Override
        void getCardFromResponse(@NonNull Map<String, AggregatedFeedContent> responses,
                                 @NonNull WikiSite wiki, int age, @NonNull List<Card> outCards) {
            for (String appLangCode : WikipediaApp.getInstance().language().getAppLanguageCodes()) {
                if (responses.containsKey(appLangCode)
                        && !FeedContentType.ON_THIS_DAY.getLangCodesDisabled().contains(appLangCode)) {
                    AggregatedFeedContent content = responses.get(appLangCode);
                    if (content.onthisday() != null && !content.onthisday().isEmpty()) {
                        outCards.add(new OnThisDayCard(content.onthisday(), WikiSite.forLanguageCode(appLangCode), age));
                    }
                }
            }
        }
    }
    public static class InTheNews extends BaseClient {
        public InTheNews(@NonNull AggregatedFeedContentClient aggregatedClient) {
            super(aggregatedClient);
        }

        @Override
        void getCardFromResponse(@NonNull Map<String, AggregatedFeedContent> responses,
                                 @NonNull WikiSite wiki, int age, @NonNull List<Card> outCards) {
            // todo: remove age check when news endpoint provides dated content, T139481.
            for (String appLangCode : WikipediaApp.getInstance().language().getAppLanguageCodes()) {
                if (responses.containsKey(appLangCode)
                        && !FeedContentType.NEWS.getLangCodesDisabled().contains(appLangCode)) {
                    AggregatedFeedContent content = responses.get(appLangCode);
                    if (age == 0 && content.news() != null) {
                        outCards.add(new NewsCard(content.news(), age, WikiSite.forLanguageCode(appLangCode)));
                    }
                }
            }
        }
    }

    public static class FeaturedArticle extends BaseClient {
        public FeaturedArticle(@NonNull AggregatedFeedContentClient aggregatedClient) {
            super(aggregatedClient);
        }

        @Override
        void getCardFromResponse(@NonNull Map<String, AggregatedFeedContent> responses,
                                 @NonNull WikiSite wiki, int age, @NonNull List<Card> outCards) {
            for (String appLangCode : WikipediaApp.getInstance().language().getAppLanguageCodes()) {
                if (responses.containsKey(appLangCode)
                        && !FeedContentType.FEATURED_ARTICLE.getLangCodesDisabled().contains(appLangCode)) {
                    AggregatedFeedContent content = responses.get(appLangCode);
                    if (content.tfa() != null) {
                        outCards.add(new FeaturedArticleCard(content.tfa(), age, WikiSite.forLanguageCode(appLangCode)));
                    }
                }
            }
        }
    }

    public static class TrendingArticles extends BaseClient {
        public TrendingArticles(@NonNull AggregatedFeedContentClient aggregatedClient) {
            super(aggregatedClient);
        }

        @Override
        void getCardFromResponse(@NonNull Map<String, AggregatedFeedContent> responses,
                                 @NonNull WikiSite wiki, int age, @NonNull List<Card> outCards) {
            for (String appLangCode : WikipediaApp.getInstance().language().getAppLanguageCodes()) {
                if (responses.containsKey(appLangCode)
                        && !FeedContentType.TRENDING_ARTICLES.getLangCodesDisabled().contains(appLangCode)) {
                    AggregatedFeedContent content = responses.get(appLangCode);
                    if (content.mostRead() != null) {
                        outCards.add(new MostReadListCard(content.mostRead(), WikiSite.forLanguageCode(appLangCode)));
                    }
                }
            }
        }
    }

    public static class FeaturedImage extends BaseClient {
        public FeaturedImage(@NonNull AggregatedFeedContentClient aggregatedClient) {
            super(aggregatedClient);
        }

        @Override
        void getCardFromResponse(@NonNull Map<String, AggregatedFeedContent> responses,
                                 @NonNull WikiSite wiki, int age, @NonNull List<Card> outCards) {
            if (responses.containsKey(wiki.languageCode())) {
                AggregatedFeedContent content = responses.get(wiki.languageCode());
                if (content.potd() != null) {
                    outCards.add(new FeaturedImageCard(content.potd(), age, wiki));
                }
            }
        }
    }

    public void invalidate() {
        aggregatedResponseAge = -1;
    }

    public void cancel() {
        disposables.clear();
    }

    private abstract static class BaseClient implements FeedClient {
        @NonNull private AggregatedFeedContentClient aggregatedClient;
        @Nullable private Callback cb;
        private WikiSite wiki;
        private int age;

        BaseClient(@NonNull AggregatedFeedContentClient aggregatedClient) {
            this.aggregatedClient = aggregatedClient;
        }

        abstract void getCardFromResponse(@NonNull Map<String, AggregatedFeedContent> responses,
                                          @NonNull WikiSite wiki, int age, @NonNull List<Card> outCards);

        @Override
        public void request(@NonNull Context context, @NonNull WikiSite wiki, int age, @NonNull Callback cb) {
            this.cb = cb;
            this.age = age;
            this.wiki = wiki;
            if (aggregatedClient.aggregatedResponseAge == age
                    && aggregatedClient.aggregatedResponses.containsKey(wiki.languageCode())) {
                List<Card> cards = new ArrayList<>();
                getCardFromResponse(aggregatedClient.aggregatedResponses, wiki, age, cards);
                FeedCoordinator.postCardsToCallback(cb, cards);
            } else {
                requestAggregated();
            }
        }

        @Override
        public void cancel() {
        }

        private void requestAggregated() {
            aggregatedClient.cancel();
            UtcDate date = DateUtil.getUtcRequestDateFor(age);
            aggregatedClient.disposables.add(Observable.fromIterable(FeedContentType.getAggregatedLanguages())
                    .flatMap(lang -> ServiceFactory.getRest(WikiSite.forLanguageCode(lang)).getAggregatedFeed(date.getYear(), date.getMonth(), date.getDay()).subscribeOn(Schedulers.io()), Pair::new)
                    .toList()
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(pairList -> {
                        List<Card> cards = new ArrayList<>();
                        for (Pair<String, AggregatedFeedContent> pair : pairList) {
                            AggregatedFeedContent content = pair.second;
                            if (content == null) {
                                continue;
                            }
                            aggregatedClient.aggregatedResponses.put(WikiSite.forLanguageCode(pair.first).languageCode(), content);
                            aggregatedClient.aggregatedResponseAge = age;
                        }
                        if (aggregatedClient.aggregatedResponses.containsKey(wiki.languageCode())) {
                            getCardFromResponse(aggregatedClient.aggregatedResponses, wiki, age, cards);
                        }
                        if (cb != null) {
                            FeedCoordinator.postCardsToCallback(cb, cards);
                        }
                    }, caught -> {
                        L.v(caught);
                        if (cb != null) {
                            cb.error(caught);
                        }
                    }));
        }
    }
}

File: app/src/main/java/org/wikipedia/feed/aggregated/AggregatedFeedContentClient.kt
package org.wikipedia.feed.aggregated

import android.content.Context
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.wikipedia.WikipediaApp
import org.wikipedia.dataclient.ServiceFactory
import org.wikipedia.dataclient.WikiSite
import org.wikipedia.feed.FeedContentType
import org.wikipedia.feed.dataclient.FeedClient
import org.wikipedia.feed.featured.FeaturedArticleCard
import org.wikipedia.feed.image.FeaturedImageCard
import org.wikipedia.feed.model.Card
import org.wikipedia.feed.news.NewsCard
import org.wikipedia.feed.onthisday.OnThisDay
import org.wikipedia.feed.onthisday.OnThisDayCard
import org.wikipedia.feed.topread.TopRead
import org.wikipedia.feed.topread.TopReadListCard
import org.wikipedia.util.DateUtil
import org.wikipedia.util.L10nUtil
import org.wikipedia.util.log.L

class AggregatedFeedContentClient {
    private val aggregatedResponses = mutableMapOf<String, AggregatedFeedContent>()
    private var aggregatedResponseAge = -1
    var clientJob: Job? = null

    class OnThisDayFeed(coroutineScope: CoroutineScope, aggregatedClient: AggregatedFeedContentClient) : BaseClient(coroutineScope, aggregatedClient) {
        override fun getCardFromResponse(responses: Map<String, AggregatedFeedContent>,
                                         wiki: WikiSite,
                                         age: Int,
                                         outCards: MutableList<Card>) {
            for (appLangCode in WikipediaApp.instance.languageState.appLanguageCodes) {
                if (responses.containsKey(appLangCode) && !FeedContentType.ON_THIS_DAY.langCodesDisabled.contains(appLangCode)) {
                    responses[appLangCode]?.randomOnThisDayEvent?.let {
                        outCards.add(OnThisDayCard(it, WikiSite.forLanguageCode(appLangCode), age))
                    }
                }
            }
        }
    }

    class InTheNews(coroutineScope: CoroutineScope, aggregatedClient: AggregatedFeedContentClient) : BaseClient(coroutineScope, aggregatedClient) {
        override fun getCardFromResponse(responses: Map<String, AggregatedFeedContent>,
                                         wiki: WikiSite,
                                         age: Int,
                                         outCards: MutableList<Card>) {
            for (appLangCode in WikipediaApp.instance.languageState.appLanguageCodes) {
                if (responses.containsKey(appLangCode) && !FeedContentType.NEWS.langCodesDisabled.contains(appLangCode)) {
                    responses[appLangCode]?.news?.let {
                        outCards.add(NewsCard(it, age, WikiSite.forLanguageCode(appLangCode)))
                    }
                }
            }
        }
    }

    class FeaturedArticle(coroutineScope: CoroutineScope, aggregatedClient: AggregatedFeedContentClient) : BaseClient(coroutineScope, aggregatedClient) {
        override fun getCardFromResponse(responses: Map<String, AggregatedFeedContent>,
                                         wiki: WikiSite,
                                         age: Int,
                                         outCards: MutableList<Card>) {
            for (appLangCode in WikipediaApp.instance.languageState.appLanguageCodes) {
                if (responses.containsKey(appLangCode) && !FeedContentType.FEATURED_ARTICLE.langCodesDisabled.contains(appLangCode)) {
                    responses[appLangCode]?.tfa?.let {
                        outCards.add(FeaturedArticleCard(it, age, WikiSite.forLanguageCode(appLangCode)))
                    }
                }
            }
        }
    }

    class TopReadArticles(coroutineScope: CoroutineScope, aggregatedClient: AggregatedFeedContentClient) : BaseClient(coroutineScope, aggregatedClient) {
        override fun getCardFromResponse(responses: Map<String, AggregatedFeedContent>,
                                         wiki: WikiSite,
                                         age: Int,
                                         outCards: MutableList<Card>) {
            for (appLangCode in WikipediaApp.instance.languageState.appLanguageCodes) {
                if (responses.containsKey(appLangCode) && !FeedContentType.TOP_READ_ARTICLES.langCodesDisabled.contains(appLangCode)) {
                    responses[appLangCode]?.topRead?.let {
                        outCards.add(
                            TopReadListCard(
                                it,
                                WikiSite.forLanguageCode(appLangCode)
                            )
                        )
                    }
                }
            }
        }
    }

    class FeaturedImage(coroutineScope: CoroutineScope, aggregatedClient: AggregatedFeedContentClient) : BaseClient(coroutineScope, aggregatedClient) {
        override fun getCardFromResponse(responses: Map<String, AggregatedFeedContent>,
                                         wiki: WikiSite,
                                         age: Int,
                                         outCards: MutableList<Card>) {
            if (responses.containsKey(wiki.languageCode)) {
                responses[wiki.languageCode]?.potd?.let {
                    outCards.add(FeaturedImageCard(it, age, wiki))
                }
            }
        }
    }

    fun invalidate() {
        aggregatedResponseAge = -1
    }

    abstract class BaseClient internal constructor(
        private val coroutineScope: CoroutineScope,
        private val aggregatedClient: AggregatedFeedContentClient
    ) : FeedClient {
        private lateinit var cb: FeedClient.Callback
        private lateinit var wiki: WikiSite
        private var age = 0

        abstract fun getCardFromResponse(responses: Map<String, AggregatedFeedContent>, wiki: WikiSite, age: Int, outCards: MutableList<Card>)

        override fun request(context: Context, wiki: WikiSite, age: Int, cb: FeedClient.Callback) {
            this.cb = cb
            this.age = age
            this.wiki = wiki
            if (aggregatedClient.aggregatedResponseAge == age && aggregatedClient.aggregatedResponses.containsKey(wiki.languageCode)) {
                val cards = mutableListOf<Card>()
                getCardFromResponse(aggregatedClient.aggregatedResponses, wiki, age, cards)
                cb.success(cards)
            } else {
                requestAggregated()
            }
        }

        override fun cancel() {}

        private fun requestAggregated() {
            aggregatedClient.clientJob?.cancel()
            val date = DateUtil.getUtcRequestDateFor(age)
            aggregatedClient.clientJob = coroutineScope.launch(
                CoroutineExceptionHandler { _, caught ->
                    L.v(caught)
                    cb.error(caught)
                }
            ) {
                val cards = mutableListOf<Card>()
                val deferredResponses = WikipediaApp.instance.languageState.appLanguageCodes.map { langCode ->
                    async {
                        val wikiSite = WikiSite.forLanguageCode(langCode)
                        val hasParentLanguageCode = !WikipediaApp.instance.languageState.getDefaultLanguageCode(langCode).isNullOrEmpty()
                        var feedContentResponse = ServiceFactory.getRest(wikiSite).getFeedFeatured(date.year, date.month, date.day)

                        feedContentResponse.randomOnThisDayEvent = feedContentResponse.onthisday?.random()

                        // TODO: This is a temporary fix for T355192
                        feedContentResponse = handleLanguageVariant(feedContentResponse, wikiSite, hasParentLanguageCode)

                        aggregatedClient.aggregatedResponses[langCode] = feedContentResponse
                        aggregatedClient.aggregatedResponseAge = age
                    }
                }

                deferredResponses.awaitAll()

                if (aggregatedClient.aggregatedResponses.containsKey(wiki.languageCode)) {
                    getCardFromResponse(aggregatedClient.aggregatedResponses, wiki, age, cards)
                }
                cb.success(cards)
            }
        }

        private suspend fun handleLanguageVariant(feedContentResponse: AggregatedFeedContent,
                                                  wikiSite: WikiSite,
                                                  hasParentLanguageCode: Boolean): AggregatedFeedContent {
            return withContext(Dispatchers.IO) {
                var response = feedContentResponse
                if (hasParentLanguageCode) {
                    val tfaDeferred = feedContentResponse.tfa?.let {
                        async { L10nUtil.getPagesForLanguageVariant(listOf(it), wikiSite, shouldUpdateExtracts = true).first() }
                    }

                    val topReadDeferred = feedContentResponse.topRead?.let {
                        async { L10nUtil.getPagesForLanguageVariant(it.articles, wikiSite) }
                    }

                    // Same logic in OnThisDayCardView and we only need to send one page to the event class.
                    val randomOnThisDayPageDeferred = feedContentResponse.randomOnThisDayEvent?.pages?.find { it.thumbnailUrl != null }?.let {
                        async { L10nUtil.getPagesForLanguageVariant(listOf(it), wikiSite) }
                    }

                    val tfaResponse = tfaDeferred?.await()
                    val topReadResponse = topReadDeferred?.await()
                    val randomOnThisDayPageResponse = randomOnThisDayPageDeferred?.await()

                    response = AggregatedFeedContent(
                        tfa = tfaResponse ?: feedContentResponse.tfa,
                        news = feedContentResponse.news,
                        topRead = topReadResponse?.let {
                            TopRead(
                                feedContentResponse.topRead.date,
                                it
                            )
                        } ?: feedContentResponse.topRead,
                        potd = feedContentResponse.potd,
                        onthisday = feedContentResponse.onthisday
                    ).apply {
                        randomOnThisDayEvent = OnThisDay.Event(
                            pages = randomOnThisDayPageResponse ?: feedContentResponse.randomOnThisDayEvent?.pages ?: emptyList(),
                            text = feedContentResponse.randomOnThisDayEvent?.text ?: "",
                            year = feedContentResponse.randomOnThisDayEvent?.year ?: 0
                        )
                    }
                }
                response
            }
        }
    }
}
